bump_rpm.sh

```diff
@@ -79,13 +79,22 @@ if [[ $CURRENT_VERSION != $NEW_VERSION ]] ; then
 	git add $SPEC_FILE
 
 	if [[ $PACKAGE_NAME == *rubygem-* ]]; then
+		CHANGELOG=$(mktemp)
+		sed -n '/%changelog/,$p' $SPEC_FILE > $CHANGELOG
+
 		TEMPLATE="$(awk '/^# template: / { print $3 }' $SPEC_FILE)"
-		if [[ $TEMPLATE == 'scl' || $TEMPLATE == 'nonscl' ]] ; then
+		if [[ $TEMPLATE == 'scl' || $TEMPLATE == 'nonscl' || $TEMPLATE == '' ]] ; then
 			TEMPLATE=default
 		fi
+
 		if [[ -n $TEMPLATE ]] ; then
-			echo "* Updating requirements"
-			gem2rpm -t $ROOT/gem2rpm/$TEMPLATE.spec.erb *.gem | $SCRIPT_DIR/update-requirements specfile - $SPEC_FILE
+			sed -i 's/\s\+$//' $SPEC_FILE
+			gem2rpm -t $ROOT/gem2rpm/$TEMPLATE.spec.erb -o $SPEC_FILE *.gem
+
+			cat $CHANGELOG >> $SPEC_FILE
+			sed -i '/^%changelog/,/^%changelog/{0,//!d}' $SPEC_FILE
+			rm $CHANGELOG
```

Comment:
This unconditionally always rewrites the entire spec, but that wipes out local changes. I think this should be hidden behind an env var.

A safer way could be that it only rewrites if the template is `scl`, `nonscl` or missing, otherwise falls back to the safer replacement.

Reply:
What scenario do you forsee where you don't want the spec file getting updated? or that you have local changes where you want to bump the RPM?

Reply:
It is mostly about the spec files where the template is insufficient (https://github.com/theforeman/foreman-packaging/pull/8166 comes to mind). I don't think we'll ever capture everything and we do have a bot now that submits PRs using the script. So I'd stay on the safe side.

Reply:
> What is the end goal there?

I'd say that it allows a conservative update by default but you can force to fully regenerate specs (on the next update) by increasing the template's version.

> Sorry, I know you said this earlier but I realize I do not know what you mean by this.

Essentially:
```sh
if [[ $TEMPLATE == 'scl' || $TEMPLATE == 'nonscl' || -z $TEMPLATE ]] ; then
  regenerate_spec_file
elif [[ -n $TEMPLATE ]] ; then
  update_spec_file
fi
```

Reply:
What is the difference between regenerate and update in this context? Are you implying that we would have something like:

```
gem2rpm/default-latest.spec.erb
gem2rpm/default-2.spec.erb
gem2rpm/default-1.spec.erb
```

And if a spec file had `# template: default 1` the regenerate action would use this template and the update action would use the latest template?

I am getting a bit lost in the nuance here and maybe that is a sign that we are overengineering things a bit. Whenever we update the version of a gem we know that there are things that may change:

 * requirements (build or runtime)
 * new files that may need to be ignored

Those things are captured by doing a specfile update.Then we have some special case scenarios where some gems may have extracurricular additions to their spec file that break the norm. My argument is this is already a problem that we run into anyway and we should incrementally start adapting to this by first accepting that we should also run a specfile update anytime we do a version bump. The script and bot do the hard work of detecting updates and opening PRs, maintainers still should review and look for anomalies and then take over the PR with updates as needed. That feels sane to me while we engineer a way to preserve the extracurricular's. That's how I looked at how we might incrementally and iteratively automate this more and more to reduce packaging burden.

Reply:
My biggest worry is that we can't rely on gem2rpm to generate a 100% correct spec, so we must make some (small) modifications. If every time the bot tries to update it those get reverted we actually get a lot of additional work. This is why I initially wrote the `update-requirements` script the way I did: allow it to replace specific parts. If we decide that files should also be replaced we can also add similar markers to `%files` to replace those bits automatically.

Then we also have cases where we do want to regenerate the entire spec file. Getting rid of the SCL macros is an excellent example: it doesn't happen often, but when it does the automation helps us a lot.

That's why I'm leaning to a cautious approach: essentially opt-in full replacement. Right now we have a case where we know it's probably the right thing to do ((non)scl to default). My versioning thoughts were about how we can do a similar thing in the future.

So short term I'd suggest:
```bash
                if [[ $TEMPLATE == 'scl' ]] || [[ $TEMPLATE == 'nonscl' ]] || [[ -z $TEMPLATE ]]; then
                        CHANGELOG=$(mktemp)
                        trap "rm -f $CHANGELOG" EXIT
                        sed -e '1,/%changelog/ d' $SPEC_FILE > $CHANGELOG
                        gem2rpm -t $ROOT/gem2rpm/default.spec.erb -o $SPEC_FILE *.gem
                        cat $CHANGELOG >> $SPEC_FILE
                        git add $SPEC_FILE
                elif [[ -n $TEMPLATE ]] ; then
                        echo "* Updating requirements"
                        gem2rpm -t $ROOT/gem2rpm/$TEMPLATE.spec.erb *.gem | $SCRIPT_DIR/update-requirements specfile - $SPEC_FILE
                        git add $SPEC_FILE
                fi
```

Then later, when we do want to force regeneration of the default template, we can think about another trigger to force regeneration.

To improve on regeneration we can change the files section to:
```erb
%files
# start specfile generated files main
%dir %{gem_instdir}
....
%{gem_spec}
# end specfile generated files main
```

And the same with docs:
```erb
<% if doc_subpackage -%>
%files doc
# start specfile generated files doc
%doc %{gem_docdir}
<%= doc_files.to_rpm %>
# end specfile generated files doc
<% end # if doc_subpackage -%>
```

This way `update-requirements` can replace files too.

Reply:
A thought I had been noodling on -- https://github.com/theforeman/foreman-packaging/pull/8169

Reply:
How about we start with https://github.com/theforeman/foreman-packaging/pull/8206 and then see where it goes from there?